rand_distr/src/negative_binomial.rs

```diff
@@ -0,0 +1,246 @@
+where
+    F: Float + FloatConst,
+    Standard: Distribution<F>,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    repr: NegativeBinomialRepr<F>,
+}
+
+/// Error type returned from `NegativeBinomial::new`.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub enum Error {
+    /// `r < 0` or `r` is `nan`
+    InvalidNumberOfSuccesses,
+
+    /// `p <= 0 || p > 1` or `p` is `nan`
+    InvalidProbability,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str(match self {
+            Error::InvalidNumberOfSuccesses => "r is NaN or not strictly positive",
+            Error::InvalidProbability => "p is NaN or outside the interval (0, 1]",
+        })
+    }
+}
+
+#[cfg(feature = "std")]
+#[cfg_attr(doc_cfg, doc(cfg(feature = "std")))]
+impl std::error::Error for Error {}
+
+impl<F> NegativeBinomial<F>
+where
+    F: Float + FloatConst,
+    Standard: Distribution<F>,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    /// Construct a new `NegativeBinomial` with the given parameters `r`
+    /// (number of successes) and `p` (probability of success on each trial).
+    pub fn new(r: F, p: F) -> Result<Self, Error> {
+        if !r.is_finite() || r <= F::zero() {
+            Err(Error::InvalidNumberOfSuccesses)
+        } else if !p.is_finite() || p <= F::zero() || p > F::one() {
+            Err(Error::InvalidProbability)
+        } else if p == F::one() {
+            Ok(NegativeBinomial {
+                repr: NegativeBinomialRepr::POne,
+            })
+        } else {
+            Ok(NegativeBinomial {
+                repr: NegativeBinomialRepr::PLessThanOne(
+                    Gamma::<F>::new(r, (F::one() - p) / p).unwrap(),
+                ),
+            })
+        }
+    }
+}
+
+impl<F> Distribution<F> for NegativeBinomial<F>
+where
+    F: Float + FloatConst,
+    Standard: Distribution<F>,
+    StandardNormal: Distribution<F>,
+    Exp1: Distribution<F>,
+    Open01: Distribution<F>,
+{
+    fn sample<R: Rng + ?Sized>(&self, rng: &mut R) -> F {
+        match &self.repr {
+            NegativeBinomialRepr::POne => F::zero(),
+            NegativeBinomialRepr::PLessThanOne(gamma) => {
+                // The method used here is to generate a Gamma(r, (1-p)/p)
+                // variate, and use it as the parameter of Poisson.  See,
+                // for example, section X.4.7 of the text *Non-Uniform Random
+                // Variate Generation* by Luc Devroye (Springer-Verlag, 1986).
+                // The gamma distribution was created in the `new()` method
+                // and saved in the NegativeBinomial instance, because it
+                // depends on just the parameters `r` and `p`.  We have to
+                // create a new Poisson instance for each variate generated.
+                Poisson::<F>::new(gamma.sample(rng)).unwrap().sample(rng)
```

Comment:
Can this happen? The Gamma distribution should return strictly positive values.

Reply:
Sorry, my point was about handling infinity as the result of simulating the gamma variable. 

Nothing ensures that a [Gamma](https://docs.rs/rand_distr/latest/rand_distr/struct.Gamma.html#impl-Distribution%3CF%3E) samples always a positive and finite float.
Not is the signature of the `sample` method nor in its documentation.

At some point there was a discussion about who should handle `infinity` out of the simulation: the library or the user. I thought the decision was that the user should handle infinity floats, maybe I am wrong. This is why I suggest handling a possible infinite value here with a loop.

Reply:
I agree, I need to fix this. If `p` is extremely small (e.g. 1e-40), then the scale passed to `Gamma` is huge (1e+40), and with such a scale, `Gamma` will generate samples that are infinity.

A simple loop would not be safe if we don't have a bound on how frequently infinity is generated.

Reply:
Yeah, I was not even thinking on extreme values, just the `unwrap`. 
The thing is, if gamma samples infinity, then the Poisson "should" also infinity.  
Then, instead of a loop, one should introduce an `if` checking if the gamma samples infinity.
If it does, return infinity directly, if it does not, sample the Poisson (created with `new` and `unwrap`).